felix/nftables/table.go

```diff
@@ -0,0 +1,1294 @@
+	if !t.chainIsReferenced(chainName) {
+		return
+	}
+	for _, r := range rules {
+		if ref, ok := r.Action.(Referrer); ok {
+			t.decrefChain(ref.ReferencedChain())
+		}
+	}
+}
+
+// increfChain increments the refcount of the given chain; if the refcount transitions from 0,
+// marks the chain dirty so it will be programmed.
+func (t *nftablesTable) increfChain(chainName string) {
+	t.logCxt.WithField("chainName", chainName).Debug("Incref chain")
+	t.chainRefCounts[chainName] += 1
+	if t.chainRefCounts[chainName] == 1 {
+		t.updateRateLimitedLog.WithField("chainName", chainName).Info("Chain became referenced, marking it for programming")
+		t.dirtyChains.Add(chainName)
+		if chain := t.chainNameToChain[chainName]; chain != nil {
+			// Recursively incref chains that this chain refers to.  If
+			// chain == nil then the chain is likely about to be added, in
+			// which case we'll handle this whe the chain is added.
+			t.maybeIncrefReferredChains(chainName, chain.Rules)
+		}
+	}
+}
+
+// decrefChain decrements the refcount of the given chain; if the refcount transitions to 0,
+// marks the chain dirty so it will be cleaned up.
+func (t *nftablesTable) decrefChain(chainName string) {
+	t.logCxt.WithField("chainName", chainName).Debug("Decref chain")
+	if t.chainRefCounts[chainName] == 1 {
+		t.updateRateLimitedLog.WithField("chainName", chainName).Info("Chain no longer referenced, marking it for removal")
+		if chain := t.chainNameToChain[chainName]; chain != nil {
+			// Recursively decref chains that this chain refers to.  If
+			// chain == nil then the chain has probably already been deleted
+			// in which case we'll already have done the decrefs.
+			t.maybeDecrefReferredChains(chainName, chain.Rules)
+		}
+		delete(t.chainRefCounts, chainName)
+		t.dirtyChains.Add(chainName)
+		return
+	}
+
+	// Chain still referenced, just decrement.
+	t.chainRefCounts[chainName] -= 1
+}
+
+func (t *nftablesTable) loadDataplaneState() {
+	// Refresh the cache of feature data.
+	t.featureDetector.RefreshFeatures()
+
+	// Load the hashes from the dataplane.
+	t.logCxt.Debug("Loading current nftables state and checking it is correct.")
+	t.opReporter.RecordOperation(fmt.Sprintf("resync-%v-v%d", t.name, t.ipVersion))
+
+	t.lastReadTime = t.timeNow()
+
+	dataplaneHashes, dataplaneRules := t.getHashesAndRulesFromDataplane()
+
+	// Check that the rules we think we've programmed are still there and mark any inconsistent
+	// chains for refresh.
+	for chainName, expectedHashes := range t.chainToDataplaneHashes {
+		logCxt := t.logCxt.WithField("chainName", chainName)
+		if t.dirtyChains.Contains(chainName) || t.dirtyBaseChains.Contains(chainName) {
+			// Already an update pending for this chain; no point in flagging it as
+			// out-of-sync.
+			logCxt.Debug("Skipping known-dirty chain")
+			continue
+		}
+		dpHashes := dataplaneHashes[chainName]
+		if !t.ourChainsRegexp.MatchString(chainName) {
```

Comment:
This block is currently not quite behaving as this is written at the moment. Namely, even the base chains are "our chains" in nftables mode, since we own the whole table.

it's still possible that some other actor might add a chain into our table, in which case we should remove it. But the implication that it's a chain we're inserting into isn't right in this mode. I think this logic could be simplified to remove the check for inserted rules, and simply always mark it as dirty.